nightguard/AppleHealthService.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  AppleHealthService.swift
+//  nightguard
+//
+//  Created by Sebastian Deisel on 02.02.22.
+//  Copyright © 2022 private. All rights reserved.
+//
+
+import Foundation
+import HealthKit
+import AVFoundation
+
+class AppleHealthService: NSObject {
+
+    static let singleton = AppleHealthService()
+
+    let healthKitStore: HKHealthStore = HKHealthStore()
+
+    private func doSync(bgData: [BloodSugar]) {
+        guard !bgData.isEmpty else { return }
+
+        let unit: HKUnit = HKUnit.init(from: UserDefaultsRepository.units.value.description)
+        let lastSyncDate: Date = UserDefaultsRepository.appleHealthLastSyncDate.value
+
+        let hkQuantitySamples: [HKQuantitySample] = bgData
+            .filter{ bloodGlucose in bloodGlucose.date > lastSyncDate }
+            .compactMap{ bloodGlucose in
+                let date: Date = bloodGlucose.date
+                let value: Double = Double(bloodGlucose.value)
+
+                return HKQuantitySample(
+                    type: getHkQuantityType(),
+                    quantity: HKQuantity(unit: unit, doubleValue: value),
+                    start: date,
+                    end: date
+                )
+            }
+
+        if (!hkQuantitySamples.isEmpty) {
+            let mostRecent: BloodSugar = bgData.max(by: { $0.date < $1.date })!
+            UserDefaultsRepository.appleHealthLastSyncDate.value = mostRecent.date
+
+            healthKitStore.save(hkQuantitySamples) { (success, error) in
+                if let error = error {
+                    print("Error saving glucose sample")
+                }
+            }
+        }
+    }
+
+    private func getHkQuantityType() -> HKQuantityType {
+        return HKQuantityType.quantityType(forIdentifier: .bloodGlucose)!
+    }
+
+    public func requestAuthorization() {
+        healthKitStore.requestAuthorization(toShare: [getHkQuantityType()], read: nil, completion:  { (success, error) in
+            return
+        })
+    }
+
+    public func isAuthorized() -> Bool {
+        return healthKitStore.authorizationStatus(for: getHkQuantityType()) == HKAuthorizationStatus.sharingAuthorized
+    }
+
+    public func sync() {
+        guard HKHealthStore.isHealthDataAvailable(),
+              isAuthorized()
+        else { return }
+
+        _ = NightscoutCacheService.singleton.loadTodaysData { [unowned self] result in
```

Comment:
hmm... Good point. I had in mind to synchronize whenever new data arrives in nightguard. Hence, I put the call in `doPeriodicUpdate()`  in `MainViewController`. I explicitly do not wanted to take a another turn and fetching data from nightscout. However, since I have `lastSyncDate` and the earliest data point from the cache, we could compare them and do an optional synchronization . I will think about that and get back to you later.

Reply:
I updated the `AppleHealthService` again. Now, the service really uses cached data from `NightscoutDataRepository`. Originally, I thought, this is transparently handled by the CacheService.

Regarding the synchronization of the uncached values: I find it difficult to integrate this behaviour into the existing infrastructure due to some edge cases. E.g. On first sync. `lastSyncDate` isn't set. Hence, it would only try to fetch the last 400 entries if I would use `NightscoutService.readChartDataWithinPeriodOfTime()` which is obviously not everything it is needed to fill the gap. I could code a relatively complex solution of course, but this would blow up the code tremendously.

Your opinion?

Reply:
Maybe I would implement it like:

- if last transmission to healthkit was on the current day, simply do it as implemented before.

If not: retrieve all the missing values. It would be something like readChartDataWithinPeriodOfTime

But omit the restriction to 400 values.

Reply:
I have a solution in my mind. I will probable implement it this evening :)

Reply:
> Maybe I would implement it like:
> 
>     * if last transmission to healthkit was on the current day, simply do it as implemented before.
> 
> 
> If not: retrieve all the missing values. It would be something like readChartDataWithinPeriodOfTime
> 
> But omit the restriction to 400 values.

The problem with this approach would be that it is unknown how many data points are missing from `lastSyncDate` to the earliest date in the cached values. So IMHO it is necessary to gradually fill the missing data until the last synchronized data point is found.

First, I wanted to simply write a loop until the termination criterion is met, however, I struggled using callbacks in these loop such that I ended up with recursive calls while hoping to not exceed the callstack size. 

A maximum of 10.000 Data points is now backfilled and synchronized with apple health.